chore: Convert Jest configs into typescript

mondoreale opened the pull request:
## Summary

I convert Jest config files into typescript for consistency.

## Changes

- Convert all jest.config.js to jest.config.ts.
- Convert associated setup and teardown files from js to ts, too.
- Changed the `ts-jest` preset to the default, which is best suited for projects written entirely in ts.
- Fix a proto-related warning so that trackerless-network build is clean (unrelated change).

## Limitations and future improvements

- n/a

## Checklist before requesting a review

- [x] Is this a breaking change? If it is, be clear in summary.
- [x] Read through code myself one more time.
- [x] Make sure any and all `TODO` comments left behind are meant to be left in.
- [x] Has reasonable passing test coverage?
- [x] Updated changelog if applicable.
- [x] Updated documentation if applicable.

mondoreale committed:
Convert jest config files to ts

mondoreale committed:
Make `trackerless-network` build without warnings #proto

teogeb commented on jest.config.ts:
```diff
@@ -0,0 +1,12 @@
+import type { Config } from '@jest/types'
+
+const config = {
+    transform: {
```
Is it intentional that these files are different? No preset and no explicit tsconfig.

mondoreale replied:
I've simply gone with defaults across the board, but I hear you, let's be explicit. The things you mention are now back in. Few notes:

- For a ts-only codebase like ours the default preset is optimal. I swapped.
- All projects w/ tests come with tsconfig.json but since they all also come with designated tsconfig.jest.json files I re-added the line.

teogeb replied:
Was the `modulePathIgnorePatterns` redundant? No big difference in the `^.+\\.ts$'` regexp, but the original was slightly better.

mondoreale replied:
Old regexp is indeed better. I'll revert. Good catch.

As for the ignored `dist`: new preset only processes ts files and I believe there are no logic-carrying ts files in dist, making ignoring it redundant.

(I've also noticed that the regexp itself excludes js. Funny.)

teogeb reviewed (changes requested):
If the content changes are intentional, could mention those in the PR description.

mondoreale committed:
Explicitly use default ts-jest preset

Best suited for projects written entirely in TypeScript.

mondoreale committed:
Point to tsconfig.jest.json

Sounds like a good idea to be explicit here, too. By default ts-jest points to tsconfig.json (most of our package tsconfig.json files extend their tsconfig.jest.json) but yeah, let's write it down openly.

teogeb commented on packages/trackerless-network/protos/NetworkRpc.proto:
```diff
@@ -3,7 +3,6 @@ syntax = "proto3";
 option optimize_for = CODE_SIZE;
 
 import "google/protobuf/empty.proto";
-import "google/protobuf/any.proto";
```
Good improvement 👍 Maybe worth mentioning in the PR description as this isn't strictly related to this PR.

mondoreale committed:
Use a better regexp

It's more precise.

mondoreale committed:
Post-rebase tweaks

Teardowns have been removed in the upstream.
